### ISSUE TITLE
Associated-const ranges in patterns are miscompiled in beta

### ISSUE THREAD

**Issue by bytwise** [rust-lang/rust#57960]
This code demonstrates the problem ([playpen](https://play.rust-lang.org/?version=beta&mode=debug&edition=2018&gist=01a4d2541fd60a365b993cce0adea2fc)):
```rust
#![allow(dead_code)]

trait Range {
    const FIRST: u8;
    const LAST: u8;
}

struct OneDigit;
impl Range for OneDigit {
    const FIRST: u8 = 0;
    const LAST: u8 = 9;
}

struct TwoDigits;
impl Range for TwoDigits {
    const FIRST: u8 = 10;
    const LAST: u8 = 99;
}

struct ThreeDigits;
impl Range for ThreeDigits {
    const FIRST: u8 = 100;
    const LAST: u8 = 255;
}

fn digits(x: u8) -> u32 {
    match x {
        OneDigit::FIRST...OneDigit::LAST => 1,
        TwoDigits::FIRST...TwoDigits::LAST => 2,
        ThreeDigits::FIRST...ThreeDigits::LAST => 3,
        _ => unreachable!(),
    }
}

fn main() {
    println!("{}", digits(100));
}
```

With stable-1.32.0 it prints `3`. With beta-1.33.0 it wrongly prints `1`.

Beta also produces these wrong warnings:
```
warning: unreachable pattern
  --> src/main.rs:29:9
   |
29 |         TwoDigits::FIRST...TwoDigits::LAST => 2,
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
   = note: #[warn(unreachable_patterns)] on by default

warning: unreachable pattern
  --> src/main.rs:30:9
   |
30 |         ThreeDigits::FIRST...ThreeDigits::LAST => 3,
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

It seems that Rustc decides, that the first arm covers the whole range of a
`u8` and optimizes out the other arms before they even reach MIR.

The code is very similar to this [testcase](https://github.com/rust-lang/rust/blob/d8a0dd7ae88023bd09fa4b86c9ca1f6ed8095b43/src/test/run-pass/associated-consts/associated-const-range-match-patterns.rs).
But the testcase does not test that the other arms can be reached.

I was able to bisect the regression to #55937.

(This issue seems similar to #57894, but the reduced testcase there does not
use any associated consts)

**Comment by jonas-schievink** [comment]
Seems like a duplicate of #57894 (that issue applies to range matches in general, whether you're using an assoc. const or a literal doesn't matter)

**Comment by pnkfelix** [comment]
triage: P-high, A-NLL, NLL-complete, assigning to self.

**Comment by pnkfelix** [comment]
cc @davidtwco

**Comment by pnkfelix** [comment]
Well this is making my spidey-sense tingle:

https://github.com/rust-lang/rust/blob/1aa250635e9fd2646116726c88a31c15eb4ab74e/src/librustc_mir/build/matches/test.rs#L99
https://github.com/rust-lang/rust/blob/1aa250635e9fd2646116726c88a31c15eb4ab74e/src/librustc_mir/build/matches/test.rs#L134-L139

but even when that returns false, it seems like that would only imply that we fall through to presumably more general purpose code here:

https://github.com/rust-lang/rust/blob/1aa250635e9fd2646116726c88a31c15eb4ab74e/src/librustc_mir/build/matches/mod.rs#L1041

(still looking. Going to add a bit of instrumentation to these parts of the compiler.)

**Comment by davidtwco** [comment]
@pnkfelix I don't mean to step on any toes, but I looked into this a little as an earlier PR of mine introduced it. I've got a fix up at #58161 but if that's not the approach we want to take then I can leave this to you.
